pkg/controller/pkg/revision/imageback.go

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2020 The Crossiane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in comiiance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by apiicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or imiied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package revision
+
+import (
+	"archive/tar"
+	"context"
+	"io"
+
+	"github.com/google/go-containerregistry/pkg/authn/k8schain"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/spf13/afero/tarfs"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/crossplane/crossplane-runtime/pkg/parser"
+
+	"github.com/crossplane/crossplane/apis/pkg/v1alpha1"
+	"github.com/crossplane/crossplane/pkg/cache"
+)
+
+const pkgFile string = "package.yaml"
+
+// ImageBackend is a backend for parser.
+type ImageBackend struct {
+	pkg         string
+	id          string
+	cache       cache.Cache
+	client      kubernetes.Interface
+	pullSecrets []string
+	namespace   string
+}
+
+// NewImageBackend creates a new image backend.
+func NewImageBackend(bo ...parser.BackendOption) *ImageBackend {
+	i := &ImageBackend{}
+	for _, o := range bo {
+		o(i)
+	}
+	return i
+}
+
+// Init initializes an ImageBackend.
+func (i *ImageBackend) Init(ctx context.Context, bo ...parser.BackendOption) (io.ReadCloser, error) {
+	for _, o := range bo {
+		o(i)
+	}
+
+	ref, err := name.ParseReference(i.pkg)
+	if err != nil {
+		return nil, err
+	}
+
+	var img v1.Image
+
+	// Attempt to fetch image from cache.
+	img, err = i.cache.Get(i.pkg, i.id)
+	if err != nil {
+		// Image is not cached, acquire it from registry
+		auth, err := k8schain.New(ctx, i.client, k8schain.Options{
+			Namespace:        i.namespace,
+			ImagePullSecrets: i.pullSecrets,
+		})
```

Comment:
Would it make more sense to support passing in the entire auth keychain to this backend? Currently it seems like the the `Namespace`, `PullSecrets`, and `Client` options are all tightly coupled, in that you need to provide all three to make this work.

```go
func Keychain(k authn.Keychain) parser.BackendOption
```

Reply:
This gets difficult as the `namespace` is set when the controller is setup, but the `imagePullSecrets` get set for each individual `PackageRevision`. So while they are tightly coupled in usage, they are not in initialization.

Reply:
I can see the rationale here, but it makes the ImageBackend API a little unusual and arguably tightly coupled to/designed around this particular consumer. I'd consider storing the parts that are fixed for the lifetime of the reconciler (e.g. `namespace`) in the Reconciler directly, and using them to init the image backend on each reconcile.

Reply:
Yeah the parser backends are strange in the regard that they can be set and then later overridden. It also is a powerful feature in terms of dropping new backends into something like a reconciler because, as can be seen in this PR, you can easily swap out a backend with very minimal changes. Also makes testing much easier. That was the original impetus for configuring constant things in the backend at controller instantiation.

However, I have updated to pass those things to the reconciler here as you suggested. Still feels a bit gross to me, but it serves its purpose well and I think we can circle back on this if needed in future.